torch/csrc/jit/script/compiler.cpp

```diff
@@ -260,97 +263,272 @@ at::Tensor getAttributeValue(const NamedValue& nv) {
+    << " but found " << total_inputs << "\n" << loc << "\n";
+    return nullptr;
   }
-  auto op = findTensorOp(n);
-  if(!op) {
-    n->destroy();
-    if(!required)
-      return nullptr;
-    throw ErrorReport(loc) << "unknown builtin op";
+  // fill in position arguments
+  for(size_t i = 0; i < inputs.size(); ++i) {
+    positional_inputs[i] = inputs[i];
   }
-  if(op->num_outputs == UNKNOWN_OUTPUTS) {
-    throw ErrorReport(loc) << "produces an unknown number of outputs, so it cannot be used directly from script methods";
+  // fill in named arguments
+  for(const NamedValue& nv : attributes) {
+    auto idx = schema.argumentIndexWithName(nv.name);
+    if(!idx) {
+      err() << "unknown keyword argument '" << nv.name << "'\n" << nv.loc;
+      return nullptr;
+    }
+    if(positional_inputs[*idx]) {
+      err() << "argument '" <<  nv.name << "' previously set \n" << nv.loc;
+      return nullptr;
+    }
+    positional_inputs[*idx] = nv;
+  }
+  // fill in default values
+  for(size_t i = 0; i < positional_inputs.size(); ++i) {
+    if(positional_inputs[i])
+      continue;
+    auto default_value = schema.arguments[i].default_value;
+    if(!default_value) {
+      err() << "argument '" << schema.arguments[i].name << "' not provided.\n" << loc;
+      return nullptr;
+    }
+    positional_inputs[i] = NamedValue(loc, i, createConstant(*method.graph(), loc, *default_value));
   }
-  for(size_t i = 0; i < op->num_outputs; ++i)
-    n->addOutput();
 
-  // special handling for the tuple that cat takes as its first argument
-  if(name == "cat") {
-    ensureTensors(loc, inputs.slice(1));
-    auto first = inputs.at(0);
-    if(first->type()->kind() != TupleType::Kind) {
-      throw ErrorReport(loc) << "expected a tuple";
+  // check input types
+  std::vector<Value*> flat_inputs;
+  for(size_t i = 0; i < schema.arguments.size(); ++i) {
+    NamedValue v = *positional_inputs[i];
+    const auto& arg = schema.arguments[i];
+
+    // implicit conversion from List[Tensor] -> Tensor for when the argument
+    // is an IntList in aten, for things like x.expand(sizes=[3,4,5])
+    if(arg.attribute_kind == AttributeKind::is &&
+       v.value->type()->isSubtypeOf(*ListType::ofTensors())) {
+      auto unpacked = graph->insertNode(graph->createTupleUnpack(v.value));
+      v.value = createStack(*graph, loc, unpacked->outputs());
     }
 
-    if(attributes.size() == 1) {
-      if(inputs.size() > 1) {
-        throw ErrorReport(loc) << "expected 1 input";
-      }
+    if(!v.value->type()->isSubtypeOf(*arg.type)) {
+      err() << "expected a value of type " << arg.type->name() << " for argument '" << arg.name << "' but found "
+            << v.value->type()->name() << "\n"
+            << v.loc;
+      return nullptr;
+    }
+
+    // we only support lists for builtins, where they must be flattened
+    if(arg.type->kind() == TypeKind::ListType) {
+      auto unpacked = graph->insertNode(graph->createTupleUnpack(v.value));
```

Comment:
TupleUnpack is a bad name now :confuse

Reply:
where is the confusion here?

Reply:
We're using tuple unpack for lists. It's generally the same as statically retrieving all elements, right?